DataFormats/Common/interface/ThinnedRefSet.h

```diff
@@ -0,0 +1,93 @@
+#ifndef DataFormats_Common_ThinnedRefSet_h
+#define DataFormats_Common_ThinnedRefSet_h
+//
+// Package:     DataFormats/Common
+// Class  :     ThinnedRefSet
+//
+/**\class ThinnedRefSet ThinnedRefSet.h "ThinnedRefSet.h"
+
+ Description: A minimal set interface (insertion, contains(), clear()) for a set of Ref keys to a thinned collection
+
+ Usage:
+
+    A ThinnedRefSet contains Ref keys to a thinned collection. The
+keys are inserted as Refs to the thinned collection itself, or to any
+parent collections of the thinned collection.
+
+The main use case are Selector classes for edm::ThinningProducer.
+There, an object of the class would be stored as a member of the
+Selector class, filled in Selector::preChoose(), calling contains() in
+Selector::choose(), and cleared in Selector::reset().
+
+Example of filling
+\code
+class ExampleSelector {
+  ...
+  edm::ThinnedRefSet<ThingCollection> keysToSave_;
+};
+
+void ExampleSelector::preChoose(edm::Handle<ThingCollection> tc, edm::Event const& e, edm::EventSetup const&) {
+  auto filler = keysToSave_.fill(edm::RefProd(tc), event.productGetter());
+  for (auto const& object : event.get(objectCollectionToken_) {
+    filler.insert(object.refToThing());
+  }
+}
```

Comment:
Just to note that _if_ the `Selector` API would be such that a return value (if any) of `preChoose()`  would be passed to `choose()`, the filling of `ThinnedRefSet` could become easier, something along
```cpp
edm::ThinnedRefSet<ThingCollection> ExampleSelector::preChoose((edm::Handle<ThingCollection> tc, edm::Event const& e, edm::EventSetup const&) {
  edm::ThinnedRefSet<ThingCollection> keysToSave(edm::RefProd(tc), event.productGetter());
  for (auto const& object : event.get(objectCollectionToken_) {
    keysToSave.insert(object.refToThing());
  }
  return keysToSave;
}

bool ExampleSelector::choose(unsigned int iIndex, Things const& iItem, edm::ThinnedRefSet<ThingCollection> const& keysToSave) const {
  return keysToSave.contains(iIndex);
}
```
An explicit `Selector::reset()` would not be needed, and this approach would also allow an `edm::global::ThinningProducer`.

(this is mostly to think out loud, I'm not planning to do this change in this PR, if ever)